File: iosApp/iosApp/ComponentViews/PredictionRowView.swift
//
//  PredictionRowView.swift
//  iosApp
//
//  Created by Simon, Emma on 3/5/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct PredictionRowView: View {
    enum PillDecoration {
        case none
        case onRow(route: Route)
        case onDirectionDestination(route: Route)
    }

    let predictions: UpcomingFormat
    let pillDecoration: PillDecoration
    let destination: () -> any View

    init(
        predictions: UpcomingFormat,
        pillDecoration: PillDecoration = .none,
        destination: @escaping () -> any View
    ) {
        self.predictions = predictions
        self.pillDecoration = pillDecoration
        self.destination = destination
    }

    var body: some View {
        HStack(spacing: 0) {
            if let secondaryAlert = predictions.secondaryAlert {
                Image(secondaryAlert.iconName)
                    .accessibilityLabel("Alert")
                    .frame(width: 18, height: 18)
                    .padding(.trailing, 8)
            }
            if case let .onRow(route) = pillDecoration {
                RoutePill(route: route, type: .flex).padding(.trailing, 8)
            }
            AnyView(destination())
            Spacer(minLength: 8)
            statuses.foregroundStyle(Color.text)
        }
        .background(Color.fill3)
        .frame(maxWidth: .infinity, minHeight: 24)
    }

    @ViewBuilder
    var statuses: some View {
        switch onEnum(of: predictions) {
        case let .some(trips):
            VStack(alignment: .trailing, spacing: 10) {
                ForEach(Array(trips.trips.enumerated()), id: \.1.id) { index, trip in
                    UpcomingTripView(
                        prediction: .some(trip.format),
                        routeType: trip.routeType,
                        isFirst: index == 0,
                        isOnly: index == 0 && trips.trips.count == 1
                    )
                }
            }
        case let .disruption(alert):
            UpcomingTripView(
                prediction: .disruption(.init(alert: alert.alert), iconName: alert.iconName),
                isFirst: true,
                isOnly: true
            )
        case let .noTrips(format):
            UpcomingTripView(prediction: .noTrips(format.noTripsFormat), isFirst: true, isOnly: true)
        case .loading:
            UpcomingTripView(prediction: .loading, isFirst: true, isOnly: true)
        }
    }
}

#Preview {
    let objects = ObjectCollectionBuilder()
    let trip1 = UpcomingTrip(trip: objects.trip { _ in })
    let trip2 = UpcomingTrip(trip: objects.trip { _ in })
    VStack(alignment: .leading, spacing: 20) {
        PredictionRowView(
            predictions: .Some(trips: [
                .init(
                    trip: trip1,
                    routeType: .commuterRail,
                    format: .ScheduleTimeWithStatusRow(
                        scheduledTime: .init(year: 2025, month: .august, day: 5, hour: 12, minute: 10, second: 0),
                        status: "Delayed",
                    ),
                    lastTrip: false,
                ),
                .init(
                    trip: trip2,
                    routeType: .commuterRail,
                    format: .Time(
                        predictionTime: .init(year: 2025, month: .august, day: 5, hour: 12, minute: 45, second: 0),
                        last: false,
                        headline: true,
                    ),
                    lastTrip: false,
                ),
            ], secondaryAlert: nil),
            destination: { Text("Needham Heights") }
        )

        PredictionRowView(
            predictions: .Some(
                trip: .init(trip: trip1, routeType: .lightRail, format: .Boarding(last: false), lastTrip: false),
                secondaryAlert: nil
            ),
            destination: { Text("Longer Destination than That") }
        )

        PredictionRowView(
            predictions: .Some(trip: .init(
                trip: trip1,
                routeType: .lightRail,
                format: .Overridden(text: "Stopped 10 stops away", last: false),
                lastTrip: false,
            ), secondaryAlert: nil),
            pillDecoration: .onRow(route: TestData.getRoute(id: "Green-B")),
            destination: { Text("Destination") }
        )

        PredictionRowView(
            predictions: .Some(trip: .init(
                trip: trip1,
                routeType: .lightRail,
                format: .Overridden(text: "Stopped 10 stops away", last: false),
                lastTrip: false,
            ), secondaryAlert: nil),
            destination: { Text("Destination") }
        )

        PredictionRowView(
            predictions: .Some(trips: [
                .init(trip: trip1, routeType: .bus, format: .ScheduleMinutes(minutes: 6, last: false), lastTrip: false),
                .init(
                    trip: trip2,
                    routeType: .bus,
                    format: .ScheduleMinutes(minutes: 15, last: false),
                    lastTrip: false
                ),
            ], secondaryAlert: nil),
            destination: { Text("Destination") }
        )

        PredictionRowView(
            predictions: .Disruption(alert: objects.alert { $0.effect = .detour }, mapStopRoute: .green),
            destination: { Text("Destination") }
        )
    }
    .padding(.leading, 16)
    .padding(.trailing, 8)
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/component/PredictionRowView.kt
package com.mbta.tid.mbta_app.android.component

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ProvideTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.MyApplicationTheme
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.generated.drawableByName
import com.mbta.tid.mbta_app.android.util.FormattedAlert
import com.mbta.tid.mbta_app.android.util.Typography
import com.mbta.tid.mbta_app.android.util.containsWrappableText
import com.mbta.tid.mbta_app.android.util.modifiers.DestinationPredictionBalance
import com.mbta.tid.mbta_app.android.util.modifiers.placeholderIfLoading
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.MapStopRoute
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder.Single.alert
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.TripInstantDisplay
import com.mbta.tid.mbta_app.model.UpcomingFormat
import com.mbta.tid.mbta_app.model.UpcomingTrip
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import kotlinx.datetime.Month

sealed interface PillDecoration {
    data class OnRow(val route: Route) : PillDecoration

    data class OnDirectionDestination(val route: Route) : PillDecoration
}

@Composable
fun PredictionRowView(
    predictions: UpcomingFormat,
    modifier: Modifier = Modifier,
    pillDecoration: PillDecoration? = null,
    destination: @Composable () -> Unit,
) {
    Row(
        modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (pillDecoration is PillDecoration.OnRow) {
            RoutePill(
                pillDecoration.route,
                line = null,
                RoutePillType.Flex,
                modifier =
                    if (predictions.secondaryAlert == null) Modifier.padding(end = 8.dp)
                    else Modifier,
            )
        }
        predictions.secondaryAlert?.let { secondaryAlert ->
            Image(
                painterResource(drawableByName(secondaryAlert.iconName)),
                stringResource(R.string.alert),
                modifier = Modifier.placeholderIfLoading().padding(end = 8.dp),
            )
        }

        Column(modifier = DestinationPredictionBalance.destinationWidth()) { destination() }
        Row(
            modifier =
                DestinationPredictionBalance.predictionWidth(predictions.containsWrappableText()),
            horizontalArrangement = Arrangement.End,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            ProvideTextStyle(value = Typography.callout) {
                Column(
                    modifier = Modifier.weight(1f, fill = false),
                    horizontalAlignment = Alignment.End,
                    verticalArrangement = Arrangement.spacedBy(4.dp),
                ) {
                    when (predictions) {
                        is UpcomingFormat.Some ->
                            predictions.trips.mapIndexed { index, prediction ->
                                Row(
                                    horizontalArrangement = Arrangement.spacedBy(2.dp),
                                    verticalAlignment = Alignment.CenterVertically,
                                ) {
                                    UpcomingTripView(
                                        UpcomingTripViewState.Some(prediction.format),
                                        modifier = Modifier.weight(1f, fill = false),
                                        routeType = prediction.routeType,
                                        isFirst = index == 0,
                                        isOnly = index == 0 && predictions.trips.count() == 1,
                                    )
                                }
                            }
                        is UpcomingFormat.Disruption ->
                            UpcomingTripView(
                                UpcomingTripViewState.Disruption(
                                    FormattedAlert(predictions.alert),
                                    iconName = predictions.iconName,
                                )
                            )
                        is UpcomingFormat.NoTrips ->
                            UpcomingTripView(
                                (UpcomingTripViewState.NoTrips(predictions.noTripsFormat))
                            )
                        is UpcomingFormat.Loading -> UpcomingTripView(UpcomingTripViewState.Loading)
                    }
                }
            }
        }
    }
}

@Preview
@Composable
private fun PredictionRowViewPreview() {
    val objects = ObjectCollectionBuilder()
    val green = objects.line()
    val greenB =
        objects.route {
            lineId = green.id.idText
            color = "00843D"
            textColor = "FFFFFF"
            shortName = "B"
            longName = "Green Line B"
            type = RouteType.LIGHT_RAIL
        }
    val trip = objects.trip()

    MyApplicationTheme {
        Column(
            Modifier.background(MaterialTheme.colorScheme.background)
                .padding(start = 16.dp, end = 8.dp)
                .padding(vertical = 10.dp),
            verticalArrangement = Arrangement.spacedBy(20.dp),
        ) {
            PredictionRowView(
                predictions =
                    UpcomingFormat.Some(
                        listOf(
                            UpcomingFormat.Some.FormattedTrip(
                                UpcomingTrip(trip),
                                RouteType.COMMUTER_RAIL,
                                TripInstantDisplay.ScheduleTimeWithStatusRow(
                                    EasternTimeInstant(2025, Month.AUGUST, 5, 12, 10),
                                    "Delayed",
                                ),
                            ),
                            UpcomingFormat.Some.FormattedTrip(
                                UpcomingTrip(trip),
                                RouteType.COMMUTER_RAIL,
                                TripInstantDisplay.Time(
                                    EasternTimeInstant(2025, Month.AUGUST, 5, 12, 45),
                                    true,
                                ),
                            ),
                        ),
                        null,
                    )
            ) {
                Text("Needham Heights")
            }

            PredictionRowView(
                predictions =
                    UpcomingFormat.Some(
                        UpcomingFormat.Some.FormattedTrip(
                            UpcomingTrip(trip),
                            RouteType.LIGHT_RAIL,
                            TripInstantDisplay.Boarding,
                        ),
                        null,
                    )
            ) {
                Text("Longer Destination than That")
            }

            PredictionRowView(
                UpcomingFormat.Some(
                    UpcomingFormat.Some.FormattedTrip(
                        UpcomingTrip(trip),
                        RouteType.LIGHT_RAIL,
                        TripInstantDisplay.Overridden("Stopped 10 stops away"),
                    ),
                    null,
                ),
                pillDecoration = PillDecoration.OnRow(greenB),
            ) {
                Text("Destination")
            }

            PredictionRowView(
                UpcomingFormat.Some(
                    UpcomingFormat.Some.FormattedTrip(
                        UpcomingTrip(trip),
                        RouteType.LIGHT_RAIL,
                        TripInstantDisplay.Overridden("Stopped 10 stops away"),
                    ),
                    null,
                )
            ) {
                Text("Destination")
            }

            PredictionRowView(
                UpcomingFormat.Some(
                    listOf(
                        UpcomingFormat.Some.FormattedTrip(
                            UpcomingTrip(trip),
                            RouteType.BUS,
                            TripInstantDisplay.ScheduleMinutes(6),
                        ),
                        UpcomingFormat.Some.FormattedTrip(
                            UpcomingTrip(trip),
                            RouteType.BUS,
                            TripInstantDisplay.ScheduleMinutes(15),
                        ),
                    ),
                    null,
                )
            ) {
                Text("Destination")
            }

            PredictionRowView(
                UpcomingFormat.Disruption(
                    alert { effect = Alert.Effect.Detour },
                    MapStopRoute.GREEN,
                )
            ) {
                Text("Destination")
            }
        }
    }
}
